flecsi/data/local/policy.hh

```diff
@@ -0,0 +1,460 @@
+    // ranks and indices of the destination of copy i.e. (local source
+    // index, {(remote dest rank, remote dest index)}). We need to do a shuffle
+    // operation to reconstruct this info from {(local ghost index, remote
+    // source rank, remote source index)}.
+    auto remote_sources =
+      destination.get_storage<const points::Value>(meta_fid);
+    // Essentially a GroupByKey of remote_sources, keys are the remote source
+    // ranks and values are vectors of remote source indices.
+    SendPoints remote_shared_entities;
+    for(const auto & [begin, end] : destination.ghost_ranges) {
+      for(auto ghost_idx = begin; ghost_idx < end; ++ghost_idx) {
+        const auto & shared = remote_sources[ghost_idx];
+        remote_shared_entities[shared.first].emplace_back(shared.second);
+        // We also group local ghost entities into
+        // (src rank, { local ghost ids})
+        ghost_entities[shared.first].emplace_back(ghost_idx);
+      }
+    }
+
+    // Create the inverse mapping of group_shared_entities. This creates a map
+    // from remote destination rank to a vector of *local* source indices. This
+    // information is later used by MPI_Send().
+    {
+      std::size_t r = 0;
+      for(auto &v : util::mpi::all_to_allv([&](int r, int) -> auto & {
+            static const std::vector<std::size_t> empty;
+            const auto i = remote_shared_entities.find(r);
+            return i == remote_shared_entities.end() ? empty : i->second;
+          })) {
+        if(!v.empty())
+          shared_entities.try_emplace(r, std::move(v));
+        ++r;
+      }
+    }
+
+    // We need to figure out the max local source index in order to give correct
+    // nelems when calling region::get_storage().
+    for(const auto & [rank, indices] : shared_entities) {
+      max_local_source_idx = std::max(max_local_source_idx,
+        *std::max_element(indices.begin(), indices.end()));
+    }
+    max_local_source_idx += 1;
+  }
+
+  // called with each field (and field_id_t) on the entity, for example, one
+  // for pressure, temperature, density etc.
+  void operator()(field_id_t data_fid) const {
+    using util::mpi::test;
+
+    // Since we are doing ghost copy via MPI, we always want the host side
+    // version. Unless we are doing some direct CUDA MPI thing in the future.
+    auto source_storage =
+      source.r->get_storage<std::byte>(data_fid, max_local_source_idx);
+    auto destination_storage = destination.get_storage<std::byte>(data_fid);
+    auto type_size = source.r->get_field_info(data_fid)->type_size;
+
+    std::vector<MPI_Request> requests;
+    requests.reserve(ghost_entities.size() + shared_entities.size());
+
+    std::vector<std::vector<std::byte>> recv_buffers;
+    for(const auto & [src_rank, ghost_indices] : ghost_entities) {
+      recv_buffers.emplace_back(ghost_indices.size() * type_size);
+      requests.resize(requests.size() + 1);
+      test(MPI_Irecv(recv_buffers.back().data(),
```

Comment:
If copy plans need `MPI_Irecv` with the HPX backend, is it meaningful to support the `!FLECSI_ENABLE_MPI` case for it at all?